src/main/java/org/opensearch/plugin/insights/core/service/categorizer/QueryShapeGenerator.java

```diff
@@ -0,0 +1,276 @@
+                    aggBuilder.getSubAggregations(),
+                    aggBuilder.getPipelineAggregations(),
+                    stringBuilder,
+                    indentCount + 2,
+                    showFields
+                );
+            }
+            aggShapeStrings.add(stringBuilder.toString());
+        }
+
+        // Sort alphanumerically and append aggregations list
+        Collections.sort(aggShapeStrings);
+        for (String shapeString : aggShapeStrings) {
+            outputBuilder.append(shapeString);
+        }
+
+        //// Pipeline Aggregation (cannot have sub-aggregations) ////
+        if (pipelineAggregations.isEmpty() == false) {
+            outputBuilder.append(baseIndent).append(TWO_SPACE_INDENT).append("pipeline aggregation:").append("\n");
+
+            List<String> pipelineAggShapeStrings = new ArrayList<>();
+            for (PipelineAggregationBuilder pipelineAgg : pipelineAggregations) {
+                pipelineAggShapeStrings.add(baseIndent + TWO_SPACE_INDENT.repeat(2) + pipelineAgg.getType() + "\n");
+            }
+
+            // Sort alphanumerically and append pipeline aggregations list
+            Collections.sort(pipelineAggShapeStrings);
+            for (String shapeString : pipelineAggShapeStrings) {
+                outputBuilder.append(shapeString);
+            }
+        }
+
+        return outputBuilder;
+    }
+
+    static String buildSortShape(List<SortBuilder<?>> sortBuilderList, Boolean showFields) {
+        if (sortBuilderList == null || sortBuilderList.isEmpty()) {
+            return "";
+        }
+        StringBuilder sortShape = new StringBuilder();
+        sortShape.append("sort:\n");
+
+        List<String> shapeStrings = new ArrayList<>();
+        for (SortBuilder<?> sortBuilder : sortBuilderList) {
+            StringBuilder stringBuilder = new StringBuilder();
+            stringBuilder.append(TWO_SPACE_INDENT).append(sortBuilder.order());
+            if (showFields) {
+                stringBuilder.append(buildFieldDataString(SORT_FIELD_DATA_MAP.get(sortBuilder.getClass()), sortBuilder));
+            }
+            shapeStrings.add(stringBuilder.toString());
+        }
+
+        Collections.sort(shapeStrings);
+        for (String line : shapeStrings) {
+            sortShape.append(line).append("\n");
+        }
+        return sortShape.toString();
+    }
+
+    static String buildFieldDataString(List<Function<Object, String>> methods, NamedWriteable builder) {
+        List<String> fieldDataList = new ArrayList<>();
+        if (methods != null) {
+            for (Function<Object, String> lambda : methods) {
+                fieldDataList.add(lambda.apply(builder));
+            }
+        }
+        return String.format(" [%s]", String.join(", ", fieldDataList));
+    }
+
+    /**
+     * Class to create field data map for query, agg, sort
+     * Map
+     */
+    public static class FieldDataMapHelper {
```

Comment:
Why are we calling this FieldData Map? It can contain other information like query range, etc going forward

Reply:
We are using field data as a general term which contains field name, type, range, etc... I can change it to something like attributes or properties but can't think of a great name atm.